File: web-client/src/api/taskApi.ts
import apiClient from './client';
import type {
  PageResponse,
  TaskParticipantRequest, TaskParticipantResponse,
  TaskResponse, TaskFullResponse, TaskSummaryResponse, TaskRequest,
  TaskCommentResponse,
  TaskProjectResponse, TaskProjectRequest,
  TaskPhaseResponse, TaskPhaseRequest, TaskPhaseUpdateRequest,
  TaskPlannedWorkRequest, TaskPlannedWorkResponse,
  TaskBookedWorkRequest, TaskBookedWorkResponse,
  TaskTimelineRequest, TaskTimelineResponse, TimelineState,
  ProjectNotificationTemplateResponse, ProjectNotificationTemplateRequest,
  TemplatePlaceholder, TaskChangeType,
  TaskAttachmentRequest, TaskAttachmentResponse, FileUploadResponse,
} from './types';

const TASKS_URL    = '/api/v1/tasks';
const PROJECTS_URL = '/api/v1/projects';
const PHASES_URL   = '/api/v1/phases';

/** Fetches a paginated list of tasks (summary view), with optional filters. */
export function getTasks(params?: { userId?: string; projectId?: string; status?: string; completionStatus?: string; page?: number; size?: number }) {
  return apiClient.get<PageResponse<TaskSummaryResponse>>(TASKS_URL, { params: { size: 20, ...params } }).then((r) => r.data);
}

/** Fetches a single task by ID. */
export function getTask(id: string) {
  return apiClient.get<TaskResponse>(`${TASKS_URL}/${id}`).then((r) => r.data);
}

/** Fetches the full task view including participants, timelines, planned work, booked work, and assigned user. */
export function getTaskFull(id: string) {
  return apiClient.get<TaskFullResponse>(`${TASKS_URL}/${id}/full`).then((r) => r.data);
}

/** Creates a new task. */
export function createTask(request: TaskRequest) {
  return apiClient.post<TaskResponse>(TASKS_URL, request).then((r) => r.data);
}

/** Updates an existing task. */
export function updateTask(id: string, request: TaskRequest) {
  return apiClient.put<TaskResponse>(`${TASKS_URL}/${id}`, request).then((r) => r.data);
}

/** Changes the phase of a task. Enforces the one-way gate: cannot return to PLANNING once left. */
export function updateTaskPhase(id: string, request: TaskPhaseUpdateRequest) {
  return apiClient.patch<TaskResponse>(`${TASKS_URL}/${id}/phase`, request).then((r) => r.data);
}

/** Soft-deletes a task. */
export function deleteTask(id: string) {
  return apiClient.delete(`${TASKS_URL}/${id}`);
}

/** Fetches all comments for a task, ordered by creation time. */
export function getTaskComments(taskId: string) {
  return apiClient.get<TaskCommentResponse[]>(`${TASKS_URL}/${taskId}/comments`).then((r) => r.data);
}

/** Adds a comment to a task and returns the created comment. */
export function addComment(taskId: string, content: string) {
  return apiClient.post<TaskCommentResponse>(`${TASKS_URL}/${taskId}/comments`, { content }).then((r) => r.data);
}

/** Fetches all participants for a task. */
export function getParticipants(taskId: string) {
  return apiClient.get<TaskParticipantResponse[]>(`${TASKS_URL}/${taskId}/participants`).then((r) => r.data);
}

/** Adds the authenticated user as a WATCHER on a task. */
export function watchTask(taskId: string) {
  return apiClient.post<TaskParticipantResponse>(`${TASKS_URL}/${taskId}/participants/watch`).then((r) => r.data);
}

/** Removes own WATCHER entry from a task. */
export function removeParticipant(taskId: string, participantId: string) {
  return apiClient.delete(`${TASKS_URL}/${taskId}/participants/${participantId}`);
}

/** Fetches all planned work entries for a task. */
export function getPlannedWork(taskId: string) {
  return apiClient.get<TaskPlannedWorkResponse[]>(`${TASKS_URL}/${taskId}/planned-work`).then((r) => r.data);
}

/** Adds a planned work entry (only allowed when task status is TODO). */
export function createPlannedWork(taskId: string, request: TaskPlannedWorkRequest) {
  return apiClient.post<TaskPlannedWorkResponse>(`${TASKS_URL}/${taskId}/planned-work`, request).then((r) => r.data);
}

/** Fetches all booked work entries for a task. */
export function getBookedWork(taskId: string) {
  return apiClient.get<TaskBookedWorkResponse[]>(`${TASKS_URL}/${taskId}/booked-work`).then((r) => r.data);
}

/** Adds a booked work entry to a task. */
export function createBookedWork(taskId: string, request: TaskBookedWorkRequest) {
  return apiClient.post<TaskBookedWorkResponse>(`${TASKS_URL}/${taskId}/booked-work`, request).then((r) => r.data);
}

/** Updates an existing booked work entry. */
export function updateBookedWork(taskId: string, bookedWorkId: string, request: TaskBookedWorkRequest) {
  return apiClient.put<TaskBookedWorkResponse>(`${TASKS_URL}/${taskId}/booked-work/${bookedWorkId}`, request).then((r) => r.data);
}

/** Soft-deletes a booked work entry. */
export function deleteBookedWork(taskId: string, bookedWorkId: string) {
  return apiClient.delete(`${TASKS_URL}/${taskId}/booked-work/${bookedWorkId}`);
}

/** Fetches all active timeline entries for a task. */
export function getTimelines(taskId: string) {
  return apiClient.get<TaskTimelineResponse[]>(`${TASKS_URL}/${taskId}/timelines`).then((r) => r.data);
}

/** Creates or updates the timeline entry for a specific state (upsert). */
export function setTimeline(taskId: string, state: TimelineState, request: TaskTimelineRequest) {
  return apiClient.put<TaskTimelineResponse>(`${TASKS_URL}/${taskId}/timelines/${state}`, request).then((r) => r.data);
}

/** Removes the timeline entry for a specific state. */
export function deleteTimeline(taskId: string, state: TimelineState) {
  return apiClient.delete(`${TASKS_URL}/${taskId}/timelines/${state}`);
}

/** Fetches all projects. */
export function getProjects() {
  return apiClient.get<TaskProjectResponse[]>(PROJECTS_URL).then((r) => r.data);
}

/** Creates a new project. */
export function createProject(request: TaskProjectRequest) {
  return apiClient.post<TaskProjectResponse>(PROJECTS_URL, request).then((r) => r.data);
}

/** Updates an existing project. */
export function updateProject(id: string, request: TaskProjectRequest) {
  return apiClient.put<TaskProjectResponse>(`${PROJECTS_URL}/${id}`, request).then((r) => r.data);
}

/** Soft-deletes a project. */
export function deleteProject(id: string) {
  return apiClient.delete(`${PROJECTS_URL}/${id}`);
}

/** Fetches all phases for a project. */
export function getPhases(projectId: string) {
  return apiClient.get<TaskPhaseResponse[]>(PHASES_URL, { params: { projectId } }).then((r) => r.data);
}

/** Creates a new phase for a project. */
export function createPhase(request: TaskPhaseRequest) {
  return apiClient.post<TaskPhaseResponse>(PHASES_URL, request).then((r) => r.data);
}

/** Updates an existing phase (e.g. sets or clears customName). */
export function updatePhase(id: string, request: TaskPhaseRequest) {
  return apiClient.put<TaskPhaseResponse>(`${PHASES_URL}/${id}`, request).then((r) => r.data);
}

/** Soft-deletes a phase. */
export function deletePhase(id: string) {
  return apiClient.delete(`${PHASES_URL}/${id}`);
}

/** Fetches all active notification templates for a project. */
export function getNotificationTemplates(projectId: string) {
  return apiClient.get<ProjectNotificationTemplateResponse[]>(`${PROJECTS_URL}/${projectId}/notification-templates`).then((r) => r.data);
}

/** Fetches the full placeholder catalogue (same for every project). */
export function getTemplatePlaceholders(projectId: string) {
  return apiClient.get<TemplatePlaceholder[]>(`${PROJECTS_URL}/${projectId}/notification-templates/placeholders`).then((r) => r.data);
}

/** Creates or replaces the notification template for a project + event type. */
export function upsertNotificationTemplate(projectId: string, eventType: TaskChangeType, request: ProjectNotificationTemplateRequest) {
  return apiClient.put<ProjectNotificationTemplateResponse>(`${PROJECTS_URL}/${projectId}/notification-templates/${eventType}`, request).then((r) => r.data);
}

/** Soft-deletes the notification template for a project + event type. */
export function deleteNotificationTemplate(projectId: string, eventType: TaskChangeType) {
  return apiClient.delete(`${PROJECTS_URL}/${projectId}/notification-templates/${eventType}`);
}

/** Uploads a file to the attachments bucket in file-service and returns its metadata. */
export function uploadFile(file: File) {
  const form = new FormData();
  form.append('file', file);
  return apiClient.post<FileUploadResponse>('/api/v1/files/attachments', form).then((r) => r.data);
}

/** Fetches all attachments for a task, ordered by upload time. */
export function getAttachments(taskId: string) {
  return apiClient.get<TaskAttachmentResponse[]>(`${TASKS_URL}/${taskId}/attachments`).then((r) => r.data);
}

/** Registers an already-uploaded file as a task attachment. */
export function addAttachment(taskId: string, request: TaskAttachmentRequest) {
  return apiClient.post<TaskAttachmentResponse>(`${TASKS_URL}/${taskId}/attachments`, request).then((r) => r.data);
}

/** Permanently deletes the attachment record and removes the file from storage. */
export function deleteAttachment(taskId: string, attachmentId: string) {
  return apiClient.delete(`${TASKS_URL}/${taskId}/attachments/${attachmentId}`);
}

File: android/data/src/main/kotlin/com/demo/taskmanager/data/api/TaskApi.kt
package com.demo.taskmanager.data.api

import com.demo.taskmanager.data.dto.AttachmentDto
import com.demo.taskmanager.data.dto.AttachmentCreateRequest
import com.demo.taskmanager.data.dto.BookedWorkCreateRequest
import com.demo.taskmanager.data.dto.BookedWorkDto
import com.demo.taskmanager.data.dto.CommentCreateRequest
import com.demo.taskmanager.data.dto.CommentDto
import com.demo.taskmanager.data.dto.NotificationTemplateDto
import com.demo.taskmanager.data.dto.NotificationTemplateRequest
import com.demo.taskmanager.data.dto.PageDto
import com.demo.taskmanager.data.dto.ParticipantDto
import com.demo.taskmanager.data.dto.PhaseCreateRequest
import com.demo.taskmanager.data.dto.PhaseDto
import com.demo.taskmanager.data.dto.PhaseUpdateRequest
import com.demo.taskmanager.data.dto.PlannedDatesRequest
import com.demo.taskmanager.data.dto.PlannedWorkDto
import com.demo.taskmanager.data.dto.TaskPhaseUpdateRequest
import com.demo.taskmanager.data.dto.ProjectCreateRequest
import com.demo.taskmanager.data.dto.ProjectDto
import com.demo.taskmanager.data.dto.TaskCreateRequest
import com.demo.taskmanager.data.dto.TaskDto
import com.demo.taskmanager.data.dto.TaskFullDto
import com.demo.taskmanager.data.dto.TaskSummaryDto
import com.demo.taskmanager.data.dto.TaskUpdateRequest
import com.demo.taskmanager.data.dto.TimelineDto
import com.demo.taskmanager.data.dto.TimelineCreateRequest
import com.demo.taskmanager.data.dto.WorkCreateRequest
import com.demo.taskmanager.data.dto.enums.TaskChangeType
import com.demo.taskmanager.data.dto.enums.TaskCompletionStatus
import com.demo.taskmanager.data.dto.enums.TaskStatus
import com.demo.taskmanager.data.dto.enums.TimelineState
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

/** Retrofit interface for task-service: projects, phases, tasks, and all task sub-resources. */
interface TaskApi {

    // ── Projects ──────────────────────────────────────────────────────────────

    @GET("api/v1/projects")
    suspend fun getProjects(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 50,
        @Query("sort") sort: String = "name",
    ): PageDto<ProjectDto>

    @GET("api/v1/projects/{id}")
    suspend fun getProject(@Path("id") id: String): ProjectDto

    @POST("api/v1/projects")
    suspend fun createProject(@Body request: ProjectCreateRequest): ProjectDto

    @PUT("api/v1/projects/{id}")
    suspend fun updateProject(
        @Path("id") id: String,
        @Body request: ProjectCreateRequest,
    ): ProjectDto

    @DELETE("api/v1/projects/{id}")
    suspend fun deleteProject(@Path("id") id: String)

    // ── Phases ────────────────────────────────────────────────────────────────

    @GET("api/v1/phases")
    suspend fun getPhases(@Query("projectId") projectId: String): List<PhaseDto>

    @GET("api/v1/phases/{id}")
    suspend fun getPhase(@Path("id") id: String): PhaseDto

    @POST("api/v1/phases")
    suspend fun createPhase(@Body request: PhaseCreateRequest): PhaseDto

    @PUT("api/v1/phases/{id}")
    suspend fun updatePhase(
        @Path("id") id: String,
        @Body request: PhaseUpdateRequest,
    ): PhaseDto

    @DELETE("api/v1/phases/{id}")
    suspend fun deletePhase(@Path("id") id: String)

    // ── Tasks ─────────────────────────────────────────────────────────────────

    @GET("api/v1/tasks")
    suspend fun getTasks(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20,
        @Query("projectId") projectId: String? = null,
        @Query("userId") userId: String? = null,
        @Query("status") status: TaskStatus? = null,
        @Query("completionStatus") completionStatus: TaskCompletionStatus? = null,
    ): PageDto<TaskSummaryDto>

    @GET("api/v1/tasks/{id}")
    suspend fun getTask(@Path("id") id: String): TaskDto

    @GET("api/v1/tasks/{id}/full")
    suspend fun getTaskFull(@Path("id") id: String): TaskFullDto

    @POST("api/v1/tasks")
    suspend fun createTask(@Body request: TaskCreateRequest): TaskDto

    @PUT("api/v1/tasks/{id}")
    suspend fun updateTask(
        @Path("id") id: String,
        @Body request: TaskUpdateRequest,
    ): TaskDto

    @PATCH("api/v1/tasks/{id}/phase")
    suspend fun changePhase(
        @Path("id") id: String,
        @Body request: TaskPhaseUpdateRequest,
    ): TaskDto

    @PUT("api/v1/tasks/{id}/planned-dates")
    suspend fun setPlannedDates(
        @Path("id") id: String,
        @Body request: PlannedDatesRequest,
    ): TaskDto

    @DELETE("api/v1/tasks/{id}")
    suspend fun deleteTask(@Path("id") id: String)

    // ── Comments ──────────────────────────────────────────────────────────────

    @GET("api/v1/tasks/{taskId}/comments")
    suspend fun getComments(@Path("taskId") taskId: String): List<CommentDto>

    @POST("api/v1/tasks/{taskId}/comments")
    suspend fun addComment(
        @Path("taskId") taskId: String,
        @Body request: CommentCreateRequest,
    ): CommentDto

    // ── Participants ──────────────────────────────────────────────────────────

    @GET("api/v1/tasks/{taskId}/participants")
    suspend fun getParticipants(@Path("taskId") taskId: String): List<ParticipantDto>

    @POST("api/v1/tasks/{taskId}/participants/watch")
    suspend fun watchTask(@Path("taskId") taskId: String): ParticipantDto

    @POST("api/v1/tasks/{taskId}/participants/join")
    suspend fun joinTask(@Path("taskId") taskId: String): ParticipantDto

    @DELETE("api/v1/tasks/{taskId}/participants/{participantId}")
    suspend fun removeParticipant(
        @Path("taskId") taskId: String,
        @Path("participantId") participantId: String,
    )

    // ── Planned work ──────────────────────────────────────────────────────────

    @GET("api/v1/tasks/{taskId}/planned-work")
    suspend fun getPlannedWork(@Path("taskId") taskId: String): List<PlannedWorkDto>

    @POST("api/v1/tasks/{taskId}/planned-work")
    suspend fun addPlannedWork(
        @Path("taskId") taskId: String,
        @Body request: WorkCreateRequest,
    ): PlannedWorkDto

    // ── Booked work ───────────────────────────────────────────────────────────

    @GET("api/v1/tasks/{taskId}/booked-work")
    suspend fun getBookedWork(@Path("taskId") taskId: String): List<BookedWorkDto>

    @POST("api/v1/tasks/{taskId}/booked-work")
    suspend fun addBookedWork(
        @Path("taskId") taskId: String,
        @Body request: BookedWorkCreateRequest,
    ): BookedWorkDto

    @PUT("api/v1/tasks/{taskId}/booked-work/{id}")
    suspend fun updateBookedWork(
        @Path("taskId") taskId: String,
        @Path("id") id: String,
        @Body request: BookedWorkCreateRequest,
    ): BookedWorkDto

    @DELETE("api/v1/tasks/{taskId}/booked-work/{id}")
    suspend fun deleteBookedWork(
        @Path("taskId") taskId: String,
        @Path("id") id: String,
    )

    // ── Timelines ─────────────────────────────────────────────────────────────

    @GET("api/v1/tasks/{taskId}/timelines")
    suspend fun getTimelines(@Path("taskId") taskId: String): List<TimelineDto>

    @PUT("api/v1/tasks/{taskId}/timelines/{state}")
    suspend fun setTimeline(
        @Path("taskId") taskId: String,
        @Path("state") state: TimelineState,
        @Body request: TimelineCreateRequest,
    ): TimelineDto

    @DELETE("api/v1/tasks/{taskId}/timelines/{state}")
    suspend fun deleteTimeline(
        @Path("taskId") taskId: String,
        @Path("state") state: TimelineState,
    )

    // ── Attachments ───────────────────────────────────────────────────────────

    @GET("api/v1/tasks/{taskId}/attachments")
    suspend fun getAttachments(@Path("taskId") taskId: String): List<AttachmentDto>

    @POST("api/v1/tasks/{taskId}/attachments")
    suspend fun addAttachment(
        @Path("taskId") taskId: String,
        @Body request: AttachmentCreateRequest,
    ): AttachmentDto

    @DELETE("api/v1/tasks/{taskId}/attachments/{attachmentId}")
    suspend fun deleteAttachment(
        @Path("taskId") taskId: String,
        @Path("attachmentId") attachmentId: String,
    )

    // ── Notification templates ─────────────────────────────────────────────────

    @GET("api/v1/projects/{projectId}/notification-templates/placeholders")
    suspend fun getTemplatePlaceholders(@Path("projectId") projectId: String): List<Map<String, String>>

    @GET("api/v1/projects/{projectId}/notification-templates")
    suspend fun getNotificationTemplates(@Path("projectId") projectId: String): List<NotificationTemplateDto>

    @GET("api/v1/projects/{projectId}/notification-templates/{eventType}")
    suspend fun getNotificationTemplate(
        @Path("projectId") projectId: String,
        @Path("eventType") eventType: TaskChangeType,
    ): NotificationTemplateDto

    @PUT("api/v1/projects/{projectId}/notification-templates/{eventType}")
    suspend fun upsertNotificationTemplate(
        @Path("projectId") projectId: String,
        @Path("eventType") eventType: TaskChangeType,
        @Body request: NotificationTemplateRequest,
    ): NotificationTemplateDto

    @DELETE("api/v1/projects/{projectId}/notification-templates/{eventType}")
    suspend fun deleteNotificationTemplate(
        @Path("projectId") projectId: String,
        @Path("eventType") eventType: TaskChangeType,
    )
}
